L1Trigger/L1THGCal/src/concentrator/HGCalConcentratorSuperTriggerCellImpl.cc

```diff
@@ -1,110 +1,163 @@
+        continue;
+      }
 
-      int TC_12th = 0;
-      int Uprime = 0;
-      int Vprime = 0;
-      int rocnum = detIdToROC_.getROCNumber(TC_idV9.triggerCellU(), TC_idV9.triggerCellV(), 1);
+      DetId tc_Id(id);
+
+      //To guard against the case in v8 geometry where
+      //there might be different thicknesses within a module
+      //This is a small effect, but in principle energy might
+      //be lost.
+      if (superTCmapping_.getCoarseTriggerCellId(id) != s.second.getSTCId()) {
+        if (triggerTools_.getTriggerGeometry()->isV9Geometry()) {
+          throw cms::Exception("NonExistingCoarseTC")
+              << "The coarse trigger cell correponsing to the nominal trigger cell does not exist";
+        } else {
+          continue;
+        }
+      }
 
-      if (rocnum == 1) {
-        Uprime = TC_idV9.triggerCellU();
-        Vprime = TC_idV9.triggerCellV() - TC_idV9.triggerCellU();
+      trigCellVecOutput.push_back(triggerCell);
 
-      } else if (rocnum == 2) {
-        Uprime = TC_idV9.triggerCellU() - TC_idV9.triggerCellV() - 1;
-        Vprime = TC_idV9.triggerCellV();
+      if (energyDivisionType_ == oneBitFraction) {  //Get the 1 bit fractions
 
-      } else if (rocnum == 3) {
-        Uprime = TC_idV9.triggerCellU() - kRotate4_;
-        Vprime = TC_idV9.triggerCellV() - kRotate4_;
+        if (id != s.second.getMaxId()) {
+          float tc_fraction = getTriggerCellOneBitFraction(s.second.getTCpt(id), s.second.getSumPt());
+          s.second.addToFractionSum(tc_fraction);
+        }
       }
+    }
+  }
 
-      TC_12th = (rocnum << kRocShift_) | ((Uprime << kUShift_ | Vprime) & kSplit_v9_);
-
-      int TC_split = TC_12th;
-      if (stcSize_.at(thickness) == kSTCsizeCoarse_) {
-        TC_split = rocnum;
-      }
+  // assign energy
+  for (l1t::HGCalTriggerCell& tc : trigCellVecOutput) {
+    const auto& stc = STCs[superTCmapping_.getCoarseTriggerCellId(tc.detId())];
+    assignSuperTriggerCellEnergyAndPosition(tc, stc);
+  }
+}
 
-      return TC_wafer << kWafer_offset_ | TC_split;
+void HGCalConcentratorSuperTriggerCellImpl::assignSuperTriggerCellEnergyAndPosition(l1t::HGCalTriggerCell& c,
+                                                                                    const SuperTriggerCell& stc) const {
+  if (energyDivisionType_ == superTriggerCell) {
+    if (c.detId() == stc.getMaxId()) {
+      c.setHwPt(stc.getSumHwPt());
+      c.setMipPt(stc.getSumMipPt());
+      c.setPt(stc.getSumPt());
+    } else {
+      throw cms::Exception("NonMaxIdSuperTriggerCell")
+          << "Trigger Cell with detId not equal to the maximum of the superTriggerCell found";
+    }
+  } else if (energyDivisionType_ == equalShare) {
+    double denominator = fixedDataSizePerHGCROC_
+                             ? double(kTriggerCellsForDivision_)
+                             : double(superTCmapping_.getConstituentTriggerCells(stc.getSTCId()).size());
+
+    c.setHwPt(stc.getSumHwPt() / denominator);
```

Comment:
slightly faster to store `denominatorInv = 1./denominator;` and then multiply by `denominatorInv` (avoiding repeated divisions)